### PR TITLE
Build protobuf as ExternalProject when not found

### PR DESCRIPTION
**Pull request by mellery451**
this replaces the ancient protobuf subtree (used only for windows builds) with a CMake ExternalProject for the latest protobuf release. Other platforms can optionally use the local build of protobuf by setting `local_protobuf=ON`

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1199,35 +1202,102 @@ target_link_libraries (ripple_libs INTERFACE NIH::nudb)
...
+if (local_protobuf OR NOT TARGET protobuf::libprotobuf)
+  message (STATUS "using local protobuf build.")
+  if (WIN32)
+    # protobuf prepends lib even on windows
+    set (lib_pre "lib")
+  else ()
+    set (lib_pre ${ep_lib_prefix})
+  endif ()
+  # for the external project build of protobuf, we currently ignore the
+  # static option and always build static libs here. This is consisten
```
Comment (jwbusch) [inline CMakeLists.txt]: typo: consistent

File: CMakeLists.txt [hunk 2]
```diff
@@ -135,6 +135,8 @@ else ()
 endif ()
 option (jemalloc "Enables jemalloc for heap profiling" OFF)
 option (werr "treat warnings as errors" OFF)
+option (local_protobuf
```
Comment (jwbusch) [inline CMakeLists.txt]: force_local_protobuf? local_protobuf seems like something discovered during config vs. an option.

I like options that have a verb in them. "FORCE_USE_LOCAL_PROTOBUF" "USE_JEMALLOC" "USE_WERR" etc.

File: CMakeLists.txt [hunk 3]
```diff
@@ -1199,35 +1202,102 @@ target_link_libraries (ripple_libs INTERFACE NIH::nudb)
...
-    list (GET status 1 status_string)
-    if (NOT status_code EQUAL 0)
-      message (FATAL_ERROR "Failed to download protoc: ${status_string}")
-    endif ()
-    set (my_protoc ${CMAKE_BINARY_DIR}/tools/protoc.exe)
+if (local_protobuf OR NOT TARGET protobuf::libprotobuf)
+  message (STATUS "using local protobuf build.")
+  if (WIN32)
+    # protobuf prepends lib even on windows
+    set (lib_pre "lib")
```
Comment (jwbusch) [inline CMakeLists.txt]: nit: maybe protobuf_lib_pre because it's a specific quirk to protobuf

File: CMakeLists.txt (hunk 1 above)
Comment (mellery451) [reply CMakeLists.txt]: fixed

File: CMakeLists.txt (hunk 2 above)
Comment (mellery451) [reply CMakeLists.txt]: I mostly prefer shorter option names just to keep configure commands from getting too crazy, but I see your point. I'm going to keep this as is for now, but if we want to take a swipe at renaming our various configure options to make them more verb-oriented in the future, that would be fine with me.

File: CMakeLists.txt (hunk 3 above)
Comment (mellery451) [reply CMakeLists.txt]: good idea - changed here as well as in the soci EP

File: CMakeLists.txt [hunk 4]
```diff
@@ -1199,35 +1202,102 @@ target_link_libraries (ripple_libs INTERFACE NIH::nudb)
...
+    IMPORTED_LOCATION_RELEASE
+      ${BINARY_DIR}/${pbuf_lib_pre}protobuf${ep_lib_suffix}
+    INTERFACE_INCLUDE_DIRECTORIES
+      ${SOURCE_DIR}/src)
+  add_dependencies (protobuf::libprotobuf protobuf_src)
+  exclude_if_included (protobuf_src)
+  exclude_if_included (protobuf::libprotobuf)
+
+  if (NOT TARGET protobuf::protoc)
+    add_executable (protobuf::protoc IMPORTED)
```
Comment (ximinez) [inline CMakeLists.txt]: Probably want to add `exclude_if_included(protobuf::protoc)` somewhere around here.

**Pull request closed without merging**